File: engine/resources.py
"""
Resource system for Catan - distribution and bank management.

Handles:
- Resource distribution on dice rolls
- Bank limits and constraints
"""

# Standard Library Imports
from typing import Tuple

# Third Party Imports
import numpy as np

# Local Imports
from .state import GameState, RESOURCE_DESERT


def distribute_resources(state: GameState, dice_roll: int) -> np.ndarray:
    """
    Distribute resources based on dice roll.

    Args:
        state: Current game state
        dice_roll: Sum of two dice (2-12)

    Returns:
        Array of shape (num_players, 5) with resources gained per player
    """
    if dice_roll == 7:
        # No resources on 7 (robber is handled separately)
        return np.zeros((state.num_players, 5), dtype=np.int16)

    topology = state.topology
    gained = np.zeros((state.num_players, 5), dtype=np.int16)

    # Find all tiles with matching number
    matching_tiles = np.where(state.tile_numbers == dice_roll)[0]

    for tile_idx in matching_tiles:
        # Skip if robber is on this tile
        if tile_idx == state.robber_tile:
            continue

        resource_type = state.tile_resources[tile_idx]
        if resource_type == RESOURCE_DESERT:
            continue

        # Check all vertices of this tile
        for vertex in topology.tile_vertices[tile_idx]:
            val = state.vertices[vertex]
            if val < 0:
                continue

            player = val % 4
            amount = 2 if val >= 4 else 1  # City gives 2, settlement gives 1

            # Check bank availability
            available = state.bank[resource_type]
            amount = min(amount, available)

            if amount > 0:
                gained[player, resource_type] += amount

    # Apply gains (respecting bank limits)
    for resource_type in range(5):
        total_needed = np.sum(gained[:, resource_type])
        available = state.bank[resource_type]

        if total_needed > available:
            # Proportionally reduce (or just give nothing per Catan rules)
            # Standard Catan: if bank runs out, nobody gets that resource
            gained[:, resource_type] = 0
        else:
            state.bank[resource_type] -= total_needed
            state.resources[: state.num_players, resource_type] += gained[
                : state.num_players, resource_type
            ]

    return gained


def roll_dice(state: GameState) -> Tuple[int, int, int]:
    """
    Roll two dice.

    Returns:
        Tuple of (die1, die2, total)
    """
    die1 = int(state.rng.integers(1, 7))
    die2 = int(state.rng.integers(1, 7))
    return die1, die2, die1 + die2


def can_afford(state: GameState, player: int, cost: np.ndarray) -> bool:
    """Check if player can afford a cost."""
    return np.all(state.resources[player] >= cost)


def pay_cost(state: GameState, player: int, cost: np.ndarray) -> bool:
    """
    Deduct cost from player's resources.

    Returns:
        True if successful
    """
    if not can_afford(state, player, cost):
        return False

    state.resources[player] -= cost
    state.bank += cost
    return True


def give_resources(state: GameState, player: int, resources: np.ndarray) -> bool:
    """
    Give resources to a player from the bank.

    Returns:
        True if successful (bank had enough)
    """
    if not np.all(state.bank >= resources):
        return False

    state.bank -= resources
    state.resources[player] += resources
    return True


def transfer_resources(
    state: GameState, from_player: int, to_player: int, resources: np.ndarray
) -> bool:
    """
    Transfer resources between players.

    Returns:
        True if successful
    """
    if not np.all(state.resources[from_player] >= resources):
        return False

    state.resources[from_player] -= resources
    state.resources[to_player] += resources
    return True

File: rust/catan-core/src/resources.rs
//! Resource distribution and bank management.

use rand::Rng;

use crate::board::{topology, RESOURCE_DESERT};
use crate::state::{GameState, MAX_PLAYERS, NUM_RESOURCES};

/// Distribute resources for a dice roll. Mirrors the Python engine exactly:
/// per-vertex gains are capped at the (pre-roll) bank level, and if a
/// resource's total demand exceeds the bank, nobody receives that resource.
pub fn distribute_resources(state: &mut GameState, dice_roll: u8) {
    if dice_roll == 7 {
        return;
    }
    let topo = topology();
    let mut gained = [[0i16; NUM_RESOURCES]; MAX_PLAYERS];

    for &tile in &state.tiles_by_number[dice_roll as usize] {
        if tile < 0 {
            break;
        }
        let tile = tile as usize;
        if tile == state.robber_tile as usize {
            continue;
        }
        let res = state.tile_resources[tile];
        if res == RESOURCE_DESERT {
            continue;
        }
        let res = res as usize;
        for &vertex in &topo.tile_vertices[tile] {
            let val = state.vertices[vertex as usize];
            if val < 0 {
                continue;
            }
            let player = (val % 4) as usize;
            let amount: i16 = if val >= 4 { 2 } else { 1 };
            let amount = amount.min(state.bank[res]);
            if amount > 0 {
                gained[player][res] += amount;
            }
        }
    }

    for res in 0..NUM_RESOURCES {
        let total: i16 = (0..state.num_players).map(|p| gained[p][res]).sum();
        if total > state.bank[res] {
            // Bank can't cover demand: per standard rules, nobody gets it.
            continue;
        }
        state.bank[res] -= total;
        for p in 0..state.num_players {
            state.resources[p][res] += gained[p][res];
        }
    }
}

pub fn roll_dice(state: &mut GameState) -> (u8, u8, u8) {
    let d1 = state.rng.gen_range(1..=6u8);
    let d2 = state.rng.gen_range(1..=6u8);
    (d1, d2, d1 + d2)
}

pub fn can_afford(state: &GameState, player: usize, cost: &[i16; 5]) -> bool {
    (0..5).all(|r| state.resources[player][r] >= cost[r])
}

pub fn pay_cost(state: &mut GameState, player: usize, cost: &[i16; 5]) -> bool {
    if !can_afford(state, player, cost) {
        return false;
    }
    for r in 0..5 {
        state.resources[player][r] -= cost[r];
        state.bank[r] += cost[r];
    }
    true
}
